testing-and-tdd/jasmine-testing.md

```diff
@@ -12,51 +12,128 @@ Here are links to lessons that should be completed before this lesson:
+  - install Chai
+  - require library for assertions
 
 ### Materials
 
 - [Jasmine Testing Slideshow](https://docs.google.com/presentation/d/1V9WEprK9j61V0S1lYtcJOXNTDMpziVMyUKHuMhpP1W0/edit?usp=sharing)
 
 ### Lesson
 
-Here's text about introducing something and how it works.
+#### Jasmine Test
+Jasmine test are BDD. You should write the test first then construct code to pass the test. If developers don't start with test it can be more time consuming to write them later.
 
-Build on the first information. Have apprentices guess things, do an activity, etc.
+**install jasmine**
 
-Make sure to mention these things:
+`npm install --global jasmine`
 
-- Things
-	- This is a sub-thing
-- More things
-- Even more things
-- Even more things
+**initialize project**
 
+`jasmine init`
 
-### Common Mistakes / Misconceptions
+`node node_modules/jasmine/bin/jasmine init`
+
+**create test files** in the "./spec" folder create files that end with "spec.js" so that jasmine knows which files are the test files. For example `string-test-spec.js`.
+
+**start test** with `npm test` in the root project directory after including `"test": "jasmine"` to the package.json scripts.
+
+**1st test**, `describe` is function that takes 2 arguments ("STRING", FUNCTION(){}) and multiple `it` statements. Inside describe `it` will contain the test statements. The testing happens in the line `expect(WORD.length == 4).toBe(true);`. Good test in jasmine should read like a sentence. `expect` and `toBe` must be true or the test will fail.
+
+```javascript
+describe("Testing string that",function(){
+    it("contains 4 letters", function(){
+        WORD = "word";
+        expect(WORD.length == 4).toBe(true);
+    });
+});
+```
+
+**multiple `it` statements** can use the same variables if they are declared under `describe` scope.
+
+```javascript
+describe("Testing string WORD",function(){
+	let WORD = "supertestinghasbegun" // failing from the start
+	
+    it("contains 4 letters", function(){
+        WORD = "word";
+
+        expect(WORD.length == 4).toBe(true);
+	}); 
+	
+    it("must be null",function(){
+        WORD = '';
+
+        expect(WORD == '').toBe(true);
+    })
+});
+```
+#### Mocha Test
+**install Mocha**
+
+`npm install --global mocha`
 
-List things that apprentices might not realize, might assume at first, or should avoid.
+**initialize Mocha file by file**
 
-- Example
-- Example
+The `.js` file that will be using mocha must require it at the top like this `const assert = require('assert');` and be placed in the `./test` folder so Mocha can apply the test.
+
+**how to test**
+
+`npm test` after adding `"test":"mocha"` to the package.json scripts.
+
+**1st test**
+
+`describe` and `it` functions work similar to jasmine's same named functions.  
+```javascript
+var assert = require('assert');
+
+describe('Mocha String Test', function () {
+ it('should return the exact number of characters in a string', function () {
+        assert.equal("Hello".length, 4);
```

Comment:
"Hello".length == 4 ?

Reply:
Showing what a fail looks like in Mocha but I think I should also add a comment to that so it is clear that is what's going on.